abstract_domains/lattice.py

```diff
@@ -188,10 +169,11 @@ def widening(self, other: 'Lattice'):
             return self._widening(other)
 
     def replace(self, other: 'Lattice') -> 'Lattice':
-        """Replace the current internal representation with the internal representation of another lattice element.
+        """Replace this instance with another lattice element.
 
         :param other: other lattice element
-        :return: current lattice element with replaced internal representation
+        :return: current lattice element updated to be equal to other
         """
-        self._internal = other._internal
+        self.__dict__.update(other.__dict__)
+        assert self == other
```

Comment:
yes. the assert check would actually fail without overwriting __eq__. It is anyway a good idea to overwrite __eq__ in this case. The default implementation uses object addresses to check for equality, but a lattice element is usually more perceived as a "value" which should be comparable like e.g. an integer.